PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp

```diff
@@ -534,6 +743,259 @@ void ProjectPanel::openSelectFile()
 }
 
 
+HMENU ProjectPanel::getContextMenu(HTREEITEM hTreeItem) const
+{
+	HMENU hMenu;
+	NodeType nodeType = getNodeType(hTreeItem);
+	if (nodeType == nodeType_root)
+		hMenu = _hWorkSpaceMenu;
+	else if (nodeType == nodeType_project)
+		hMenu = _hProjectMenu;
+	else if (nodeType == nodeType_folder)
+		hMenu = _hFolderMenu;
+	else if (nodeType == nodeType_baseDir)
+		hMenu = _hDirectoryMenu;
+	else if (nodeType == nodeType_dir || nodeType == nodeType_dirFile)
+		hMenu = NULL; // currently no context menu for monitored files
+	else //nodeType_file
+		hMenu = _hFileMenu;
+	return hMenu;
+
+}
+
+void ProjectPanel::onTreeItemAdded(bool afterClone, HTREEITEM hItem, TreeViewData* newData)
+{
+	afterClone;
+
+	ProjectPanelData* tvInfo = static_cast<ProjectPanelData*> (newData);
+	tvInfo->setItem( hItem );
+	if (tvInfo->isBaseDirectory())
+		tvInfo->watchDir(true);
+}
+
+void ProjectPanel::onMessage(HWND hwnd, UINT message, WPARAM wParam, LPARAM lParam)
+{
+	wParam;
+
+	switch (message)
+	{
+		case DIRECTORYWATCHER_UPDATE:
+		{
+			HTREEITEM hItem = (HTREEITEM)lParam;
+			if (_treeView.itemValid(hItem))
+			{
+
+				TVITEM tvItem;
+				tvItem.hItem = hItem;
+				tvItem.mask = TVIF_PARAM;
+				SendMessage(hwnd, TVM_GETITEM, 0,reinterpret_cast<LPARAM>(&tvItem));
+				treeItemChanged(hItem, (TreeViewData*) tvItem.lParam);
+			}
+		}
+		break;
+	}
+
+}
+
+void ProjectPanel::treeItemChanged(HTREEITEM hTreeItem, TreeViewData* tvdata)
+{
+
+	TVITEM tvItem;
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hTreeItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));

```

Comment:
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem`.